cluster_management/src/main/java/com/pinterest/rocksplicator/eventstore/LeaderEventHistoryStore.java

```diff
@@ -0,0 +1,499 @@
+
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.CLIENT_OBSERVED_SHARDMAP_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.CLIENT_OBSERVED_SHARDMAP_LEADER_UP;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_OBSERVED_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_OBSERVED_LEADER_UP;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_POSTED_SHARDMAP_LEADER_DOWN;
+import static com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType.SPECTATOR_POSTED_SHARDMAP_LEADER_UP;
+
+import com.pinterest.rocksplicator.codecs.Codec;
+import com.pinterest.rocksplicator.codecs.WrappedDataThriftCodec;
+import com.pinterest.rocksplicator.thrift.commons.io.CompressionAlgorithm;
+import com.pinterest.rocksplicator.thrift.commons.io.SerializationProtocol;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEvent;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventType;
+import com.pinterest.rocksplicator.thrift.eventhistory.LeaderEventsHistory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.math.IntMath;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+
+public class LeaderEventHistoryStore implements Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeaderEventHistoryStore.class);
+  private static final int NUM_PARALLEL_THREADS = 64;
+
+  private final String zkConnectString;
+  private final String clusterName;
+  private final CuratorFramework zkClient;
+  private final Optional<Integer> maxEventsToKeep;
+  private final Codec<LeaderEventsHistory, byte[]> leaderEventsHistoryCodec;
+  private final List<ExecutorService> executorServices;
+
+  private final LoadingCache<String, LoadingCache<String,
+      MergeableReadWriteStore<LeaderEventsHistory, LeaderEvent>>>
+      zkStoreCache;
+
+  private final LoadingCache<String, Cache<String, LeaderEventsHistory>> localHistoryCache;
```

Comment:
what is the purpose of this "locakHistoryCache", I have not seen usage of this use in this PR.

Reply:
line number 234.

It is used to prune out potentially duplicate events from locally known history of events. This helps remove any dups like in Spectator when it receives ExternalView, for a leader that never changes it's state... we could be publishing same event for same state of leader multiple times... We are not interested in such events. We are only interested in events when from an observer's perspective the state of a leader has changed.